### DISCUSSION TITLE
MailerInterface is not working

### DISCUSSION THREAD

**Discussion by harrydeboer** [symfony/symfony#57362] [Q&A] [answered]
### Symfony version(s) affected

7.0.8

### Description

I am sending mail using the MailerInterface. Sending mail gives no error but the mail is not delivered. When I am using the Mailer and Transport classes directly the mail is delivered. This behavior exists since I use Symfony 7.

### How to reproduce

.env.local
MAILER_DSN=smtp://USERNAME:APP_PASSWORD@smtp.example.com:000

`use Symfony\Component\Mailer\MailerInterface;`

`public function __construct(
private readonly MailerInterface $mailer,
) {
}`

`               $email = (new Email())
                    ->from(new Address('postmaster@example.com', strip_tags($form->get('name')->getData())))
                    ->replyTo($form->get('email')->getData())
                    ->to('info@example.com')
                    ->subject(strip_tags($form->get('subject')->getData()))
                    ->html($form->get('message')->getData());`

`$this->mailer->send($email);`

The send method of mailer gives no exception but the mail is not delivered.

### Possible Solution

When I explicitly set the transport to the mailer it mails correctly:

`use Symfony\Component\Mailer\Mailer;`
`use Symfony\Component\Mailer\Transport;`

                $transport = Transport::fromDsn($this->getParameter('mailer_dsn'));
                $mailer = new Mailer($transport);

`                $email = (new Email())
                    ->from(new Address('postmaster@example.com', strip_tags($form->get('name')->getData())))
                    ->replyTo($form->get('email')->getData())
                    ->to('info@example.com')
                    ->subject(strip_tags($form->get('subject')->getData()))
                    ->html($form->get('message')->getData());`

`$mailer->send($email);`

**Comment by xabbuh** [comment]
Can you create a small example application that allows to reproduce your issue?

**Reply by harrydeboer** [reply]
Create a new project: composer create-project symfony/skeleton:"7.0.*" project
then in project:
composer require webapp

Add the following to services.yaml
`parameters:`
`    mailer_dsn: '%env(MAILER_DSN)%'`

Make a local env file:
.env.local

APP_ENV=dev
APP_SECRET=secret
DB_USERNAME=root
DB_PASSWORD=secret
MYSQL_ROOT_PASSWORD=${DB_PASSWORD}
MYSQL_PASSWORD=${DB_PASSWORD}
DB_DATABASE=mailbug
DB_HOST=localhost
DB_CONNECTION=mysql
DB_PORT=3306
DATABASE_URL="${DB_CONNECTION}://${DB_USERNAME}:${DB_PASSWORD}@${DB_HOST}:${DB_PORT}/${DB_DATABASE}?serverVersion=8.0.37"
MAILER_DSN=smtp://info@example.com:APP-PASSWORD@smtp.gmail.com:587

From a controller I can send the email.

`namespace App\Controller;`

`use Symfony\Bundle\FrameworkBundle\Controller\AbstractController;`
`use Symfony\Component\HttpFoundation\Response;`
`use Symfony\Component\Mailer\Mailer;`
`use Symfony\Component\Mailer\MailerInterface;`
`use Symfony\Component\Mailer\Transport;`
`use Symfony\Component\Mime\Address;`
`use Symfony\Component\Mime\Email;`
`use Symfony\Component\Routing\Attribute\Route;`

`class Controller extends AbstractController`
`{`
    `public function __construct(
        private readonly MailerInterface $mailer,
    ) {
    }`

`    #[
        Route('/', name: 'homepage'),
    ]
    public function view(): Response`
`    {`
`        $email = (new Email())
            ->from(new Address('postmaster@example.com', 'John'))
            ->replyTo('john@test.com')
            ->to('info@example.com')
            ->subject('Test')
            ->html('Test');`

`        $this->mailer->send($email);`

 `       return new Response();`
`    }`
`}`

it sends no email now, but when I change the code to:

`namespace App\Controller;`

`use Symfony\Bundle\FrameworkBundle\Controller\AbstractController;`
`use Symfony\Component\HttpFoundation\Response;`
`use Symfony\Component\Mailer\Mailer;`
`use Symfony\Component\Mailer\MailerInterface;`
`use Symfony\Component\Mailer\Transport;`
`use Symfony\Component\Mime\Address;`
`use Symfony\Component\Mime\Email;`
`use Symfony\Component\Routing\Attribute\Route;`

`class Controller extends AbstractController`
`{`
    `public function __construct(
        private readonly MailerInterface $mailer,
    ) {
    }`

`    #[
        Route('/', name: 'homepage'),
    ]
    public function view(): Response`
`    {`
`        $transport = Transport::fromDsn($this->getParameter('mailer_dsn'));`
`        $mailer = new Mailer($transport);`

`        $email = (new Email())
            ->from(new Address('postmaster@example.com', 'John'))
            ->replyTo('john@test.com')
            ->to('info@example.com')
            ->subject('Test')
            ->html('Test');`

`        $mailer->send($email);`

 `       return new Response();`
`    }`
`}`

Using the Transport in the Mailer directly makes the mailing work.

**Reply by MatTheCat** [reply]
`composer require webapp` will install `symfony/messenger` so it may be that by default your emails are sent as messages to a bus.

**Reply by sfmok** [reply, marked as answer]
Based on [Mailer component](https://symfony.com/doc/current/mailer.html) doc and as @MatTheCat  mentioned:

`if your application has the Messenger component installed, all emails will be sent asynchronously by default`

This mean emails are end up as messages to the bus but you can change that https://symfony.com/doc/current/messenger.html#messenger-handling-messages-synchronously
